src/test/ui/borrowck/slice-index-bounds-check-invalidation.rs

```diff
@@ -0,0 +1,63 @@
+// Test that we error if a slice is modified after it has been bounds checked
+// and before we actually index it.
+
+fn modify_before_assert_slice_slice(x: &[&[i32]]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z[..]];
+    x[{ x = y; 0 }][2]              // OK we haven't checked any bounds before we index `x`.
+}
+
+fn modify_before_assert_array_slice(x: &[&[i32]; 3]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z[..], &z[..], &z[..]];
+    x[{ x = y; 0 }][2]              // OK we haven't checked any bounds before we index `x`.
+}
+
+fn modify_before_assert_slice_array(x: &[&[i32; 3]]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z];
+    x[{ x = y; 0 }][2]              // OK we haven't checked any bounds before we index `x`.
+}
+
+fn modify_before_assert_array_array(x: &[&[i32; 3]; 3]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z, &z, &z];
+    x[{ x = y; 0 }][2]              // OK we haven't checked any bounds before we index `x`.
+}
+
+fn modify_after_assert_slice_slice(x: &[&[i32]]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z[..]];
+    x[1][{ x = y; 2}]               //~ ERROR cannot assign `x` in indexing expression
+}
+
+fn modify_after_assert_array_slice(x: &[&[i32]; 1]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z[..]];
+    x[0][{ x = y; 2}]               // OK cannot invalidate a fixed-size array bounds check
+}
+
+fn modify_after_assert_slice_array(x: &[&[i32; 3]]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z];
+    x[1][{ x = y; 2}]               //~ ERROR cannot assign `x` in indexing expression
+}
+
+fn modify_after_assert_array_array(x: &[&[i32; 3]; 1]) -> i32 {
+    let mut x = x;
+    let z = [1, 2, 3];
+    let y = &[&z];
+    x[0][{ x = y; 2}]               // OK cannot invalidate a fixed-size array bounds check
```

Comment:
you might as well add one more set of tests checking the case `x[0][{ x = y; 0 }][2]`, right? (As in, something where the problematic mutation doesn't happen in the last indexing expression, but rather in a middle one. At least my reading of this code is that we are planning to accept mutations of the base local in the first indexing expression, but generally you're not going to be able to mutate it in any of the indexing expressions beyond the first..)